google/cloud/spanner/row_parser.h

```diff
@@ -30,42 +31,60 @@ namespace spanner {
 inline namespace SPANNER_CLIENT_NS {
 
 /**
- * A RowParser takes a range of `Value` objects and a list of types, and it
- * parses the values into a range of `Row` objects with the specified types.
+ * A functor that consumes and yields a `Value` object from some source.
+ *
+ * The returned `Value` is wrapped in an `optional` and a `StatusOr`. If there
+ * was an error getting the `Value` to return, a non-OK `Status` should be
+ * returned. If there was no error, but the source is empty, an OK `Status`
+ * with an empty `optional<Value>` should be returned.
+ *
+ * The following example shows how to create a `ValueSource` from a vector:
+ *
+ *     ValueSource MakeValueSource(std::vector<Value> const& v) {
+ *       std::size_t i = 0;
+ *       return [=]() mutable -> StatusOr<optional<Value>> {
+ *           if (i == v.size()) return optional<Value>{};
+ *           return {v[i++]};
+ *       };
+ *     }
+ */
+using ValueSource = std::function<StatusOr<optional<Value>>()>;
+
+/**
+ * A `RowParser` converts the given `ValueSource` into a single-pass iterable
+ * range of `Row<Ts...>` objects.
+ *
+ * Instances of this class are typically obtained from the
+ * `ResultSet::rows<Ts...>` member function. Callers should iterate `RowParser`
+ * using a range-for loop as follows.
  *
- * Callers should create instances of this class with the `MakeRowParser()`
- * function (defined below), which takes an input range and a list of types
- * that correspond to each row. The returned `RowParser` can be iterated,
- * producing a range of `StatusOr<Row<Ts...>>`. The `StatusOr` is needed in
- * case the input `Value` objects cannot be parsed to the requested type.
  *
  * Example:
  *
- *     std::vector<Value> const values = {
- *         Value(true), Value(0),  // Row 0
- *         Value(true), Value(1),  // Row 1
- *         Value(true), Value(2),  // Row 2
- *         Value(true), Value(3),  // Row 3
- *     };
- *     for (auto row : MakeRowParser<bool, std::int64_t>(values)) {
+ *     ValueSource vs = ...
+ *     RowParser<bool, std::int64_t> rp(std::move(vs));
+ *     for (auto row : rp) {
  *       if (!row) {
  *         // handle error
```

Comment:
Should we explicitly `continue` (or perhaps `break`) here?

Reply:
a `break` sounds right.